torch/csrc/distributed/c10d/init.cpp

```diff
@@ -700,7 +701,32 @@ They are used in specifying strategies for reduction collectives, e.g.,
           "wait",
           &::c10d::ProcessGroup::Work::wait,
           py::arg("timeout") = kNoTimeout,
-          py::call_guard<py::gil_scoped_release>());
+          py::call_guard<py::gil_scoped_release>())
+      .def(
+          "get_future",
+          [](::c10d::ProcessGroup::Work& work)
+              -> std::shared_ptr<jit::PythonFutureWrapper> {
+            return std::make_shared<jit::PythonFutureWrapper>(work.getFuture());
+          },
+          py::call_guard<py::gil_scoped_release>(),
+          R"(
+            ``get_future`` retrieves a future associated with the completion of
+            ``c10d.ProcessGroup.work``. As an example, a future object can be
+            retrieved by `fut = dist.allreduce(tensors).get_future()`.
+
+            In the example above, if `dist.allreduce` work was done on GPU,
+            `fut.wait()` would return after synchronizing the correct GPU streams
+            to ensure we can have async CUDA execution and it does not wait for
+            the entire operation to complete on GPU. If NCCL_BLOCKING_WAIT is enabled,
+            in that case, it would wait for the entire operation to complete before
```

Comment:
Let's add an example below and use code + comments to explain how `NCCL_BLOCKING_WAIT` changes the behavior.

Reply:
Based on the internal discussion, I think it is okay to leave this as is. just in case cc'ing @pritamdamania87